go/controller/internal/controller/servicedeployment_controller.go

```diff
@@ -496,57 +497,59 @@ func (r *ServiceReconciler) addOwnerReferences(ctx context.Context, service *v1a
-				controllerutil.RemoveFinalizer(service, ServiceFinalizer)
-				return ctrl.Result{}, nil
-			}
-			utils.MarkCondition(service.SetCondition, v1alpha1.SynchronizedConditionType, v1.ConditionFalse, v1alpha1.SynchronizedConditionReasonError, err.Error())
-			return ctrl.Result{}, err
+func (r *ServiceReconciler) addOrRemoveFinalizer(service *v1alpha1.ServiceDeployment) *ctrl.Result {
+	// If the service is not being deleted and if it does not have the finalizer, then let's add it.
+	if service.ObjectMeta.DeletionTimestamp.IsZero() && !controllerutil.ContainsFinalizer(service, ServiceFinalizer) {
+		controllerutil.AddFinalizer(service, ServiceFinalizer)
+	}
+
+	// If the service is being deleted, cleanup and remove the finalizer.
+	if !service.ObjectMeta.DeletionTimestamp.IsZero() {
+		// If the service does not have an ID, the finalizer can be removed.
+		if !service.Status.HasID() {
+			controllerutil.RemoveFinalizer(service, ServiceFinalizer)
+			return &ctrl.Result{}
 		}
 
-		existingService, err := r.ConsoleClient.GetService(*cluster.Status.ID, service.ConsoleName())
-		if err != nil && !errors.IsNotFound(err) {
-			utils.MarkCondition(service.SetCondition, v1alpha1.SynchronizedConditionType, v1.ConditionFalse, v1alpha1.SynchronizedConditionReasonError, err.Error())
-			return ctrl.Result{}, err
+		// If the service is already being deleted from Console API, requeue.
+		if r.ConsoleClient.IsServiceDeleting(service.Status.GetID()) {
+			return &waitForResources
 		}
-		if existingService != nil && existingService.DeletedAt != nil {
-			log.Info("waiting for the console")
-			updateStatus(service, existingService, "")
-			return requeue, nil
+
+		exists, err := r.ConsoleClient.IsServiceExisting(service.Status.GetID())
+		if err != nil {
+			return &requeue
 		}
-		if existingService != nil {
-			if err := r.deleteService(service); err != nil {
+
+		// Remove service from Console API if it exists and is not read-only.
+		if exists && !service.Status.IsReadonly() {
+			if err := r.deleteService(service.Status.GetID(), service.Spec.Detach); err != nil {
+				// If it fails to delete the external dependency here, return with the error
+				// so that it can be retried.
 				utils.MarkCondition(service.SetCondition, v1alpha1.SynchronizedConditionType, v1.ConditionFalse, v1alpha1.SynchronizedConditionReasonError, err.Error())
-				return ctrl.Result{}, err
+				return &ctrl.Result{}
 			}
-			return requeue, nil
+
+			// If the deletion process started requeue so that we can make sure the service
+			// has been deleted from Console API before removing the finalizer.
+			return &waitForResources
 		}
+
+		// If our finalizer is present, remove it.
 		controllerutil.RemoveFinalizer(service, ServiceFinalizer)
+
+		// Stop reconciliation as the item does no longer exist.
+		return &ctrl.Result{}
 	}
-	return ctrl.Result{}, nil
+
+	return nil
 }
 
-func (r *ServiceReconciler) deleteService(service *v1alpha1.ServiceDeployment) error {
```

Comment:
Deletion was previously based on ID from CRD, but since we get object from API it's better to use it as it will always be there.